internal/api/middleware.go

```diff
@@ -0,0 +1,80 @@
+package server
+
+import (
+	"log"
+	"net/http"
+
+	"github.com/canonical/gocert/internal/metrics"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+type middleware func(http.Handler) http.Handler
+
+// The middlewareContext type helps middleware receive and pass along information through the middleware chain.
+type middlewareContext struct {
+	responseStatusCode int
+	metrics            *metrics.PrometheusMetrics
+}
+
+// The responseWriterCloner struct wraps the http.ResponseWriter struct, and extracts the status
+// code of the response writer for the middleware to read
+type responseWriterCloner struct {
+	http.ResponseWriter
+	statusCode int
+}
```

Comment:
I'm somewhat confused by this approach. Is this common practice?

Reply:
Wrapping functions and implementing stdlib interfaces I've found to be very common in Go programs. There are some tutorials that also show this as an example for being able to read from writers. There are also a bunch of native functions for this like io.MultiWriter and io.TeeReader, but they required modifying the handlers themselves which I wanted to avoid for coupling reasons.